File: src/webapps/MetadataServicesToolkit/page-resources/js/user_registration.js
 /*
  * Copyright (c) 2009 University of Rochester
  *
  * This program is free software; you can redistribute it and/or modify it under the terms of the MIT/X11 license. The text of the  
  * license can be found at http://www.opensource.org/licenses/mit-license.php and copy of the license can be found on the project
  * website http://www.extensiblecatalog.org/. 
  *
  */
  
YAHOO.namespace("xc.mst.registeration");



YAHOO.xc.mst.registeration = {

	register : function()
	{

		if (document.getElementById('user_full_name').value=='') {
			alert('User Full Name is required.');
			return false;
		}
		if (document.getElementById('login_server').value== '0') {
			alert('Please select login sever type.');
			return false;
		}						

		if (document.getElementById('user_name').value=='') {
			alert('User name is required.');
			return false;
		}
		
		

		if (!document.getElementById('user_password').disabled) {
			if (document.getElementById('user_password').value=='') {
				alert('Password is required.');
				return false;
			}
			if (document.getElementById('user_password').value!=document.getElementById('user_password_confirmation').value) {
				alert('Password does not match confirmation password.');
				return false;
			}
			
			if (document.getElementById('user_password').value.length < 6 || document.getElementById('user_password').value.length > 20) {
				alert('Password should be 6 to 10 characters long');
				return false;
			}
					
			var alphaExp = /^[0-9a-zA-Z]+$/;
			
			var validChar = '*,(,),_,#,@';
			var i = 0;
			for(i = 0; i < document.getElementById('user_password').value.length; i++) {
				if(document.getElementById('user_password').value[i].match(alphaExp)){

				}else{
					if  (validChar.indexOf(document.getElementById('user_password').value[i]) < 0 ) {
						alert('Invalid character in password. Only A-Z, a-z, 0-9 , *, @, (,), _,# are allowed.');
						return false;
					}
				}
			}

		}
		if (document.getElementById('user_email').value=='') {
			alert('Email is required.');
			return false;
		} else {
			if (!YAHOO.xc.mst.registeration.emailCheck(document.getElementById('user_email').value)) {
				alert("Invalid E-mail ID");
				return false;
			}

		}

		document.registerationForm.action= basePath + 'registerUser.action';
		document.registerationForm.submit();
	
	},

	cancel : function()
	{
		document.registerationForm.action= basePath + 'home.action';
		document.registerationForm.submit();
	
	},
	
	emailCheck : function(str) 
	{
			var at="@"
			var dot="."
			var lat=str.indexOf(at)
			var lstr=str.length
			var ldot=str.indexOf(dot)
			if (str.indexOf(at)==-1){
			   return false
			}
	
			if (str.indexOf(at)==-1 || str.indexOf(at)==0 || str.indexOf(at)==lstr){
			   return false
			}
	
			if (str.indexOf(dot)==-1 || str.indexOf(dot)==0 || str.indexOf(dot)==lstr){
			    return false
			}
	
			 if (str.indexOf(at,(lat+1))!=-1){
			    return false
			 }
	
			 if (str.substring(lat-1,lat)==dot || str.substring(lat+1,lat+2)==dot){
			    return false
			 }
	
			 if (str.indexOf(dot,(lat+2))==-1){
			    return false
			 }
			
			 if (str.indexOf(" ")!=-1){
			    return false
			 }
	
	 		 return true					
	},
	
	/*
	 * Show the form to enter the password
	 */
	determinePasswordBoxDisplay : function() 
	{

		if (document.getElementById("login_server").value == 'Local') {
			document.getElementById("user_password").disabled = false;
			document.getElementById("user_password_confirmation").disabled = false;
		} else {
			document.getElementById("user_password").disabled = 'true';
			document.getElementById("user_password_confirmation").disabled = 'true';
		} 
	},

	init : function()
    	{
		YAHOO.xc.mst.registeration.determinePasswordBoxDisplay();
	}
};



// initialize the code once the dom is ready
YAHOO.util.Event.onDOMReady(YAHOO.xc.mst.registeration.init);

File: src/java/xc/mst/action/user/UserRegistration.java
/**
  * Copyright (c) 2009 University of Rochester
  *
  * This program is free software; you can redistribute it and/or modify it under the terms of the MIT/X11 license. The text of the
  * license can be found at http://www.opensource.org/licenses/mit-license.php and copy of the license can be found on the project
  * website http://www.extensiblecatalog.org/.
  *
  */

package xc.mst.action.user;

import java.util.List;

import org.apache.log4j.Logger;

import xc.mst.bo.user.Server;
import xc.mst.bo.user.User;
import xc.mst.constants.Constants;
import xc.mst.dao.DataException;
import xc.mst.email.Emailer;
import xc.mst.manager.user.DefaultServerService;
import xc.mst.manager.user.DefaultUserService;
import xc.mst.manager.user.ServerService;
import xc.mst.manager.user.UserService;

import com.opensymphony.xwork2.ActionSupport;

/**
 * Action to register a user
 *
 * @author Sharmila Ranganathan
 *
 */
public class UserRegistration extends ActionSupport {

	/** Generated id  */
	private static final long serialVersionUID = 5946519493525167816L;

	/** A reference to the logger for this class */
	static Logger log = Logger.getLogger(Constants.LOGGER_GENERAL);

	/** New user registering with the system */
	private User newUser;

	/** List of servers */
	private List<Server> servers;

	/** Server service */
	private ServerService serverService =  new DefaultServerService();

	/** User service */
	private UserService userService =  new DefaultUserService();

	/** Comments entered by user */
	private String comments;

	/** Name of server selected */
	private String serverName;

	/**
	 * To register the user
	 */
	public String execute() throws DataException {
		servers =  serverService.getAll();

		return SUCCESS;
	}

	/**
	 * Registers the new user
	 *
	 * @return
	 */
	public String registerUser() {

		log.debug(UserRegistration.class + ":" + "registerUser()" );
		Server server = serverService.getServerByName(serverName);
		User otherUser = userService.getUserByUserName(newUser.getUsername().trim(), server);

		try {
			if (otherUser == null) {
				User otherUserWithSameEmail = userService.getUserByEmail(newUser.getEmail().trim(), server);

				if (otherUserWithSameEmail == null) {

					newUser.setServer(server);

					boolean emailSent = false;

					// Email the user
					Emailer emailer = new Emailer();
					StringBuffer messageBody = new StringBuffer();
					messageBody.append("Account has been created successfully in Metadata Services Toolkit.\n");
					messageBody.append("You will be able to login once System admin assigns permissions for your account.");
					String subject = "Your New Metadata Services Toolkit account";

					emailSent = emailer.sendEmail(newUser.getEmail().trim(), subject, messageBody.toString());

					if (!emailSent) {
						servers =  serverService.getAll();
						StringBuffer errorMessage = new StringBuffer();
						errorMessage.append("Email verification failed. Either the Email address doesnot exist or some problem with the mail server.\n");
						errorMessage.append("So user registartion Failed. Please enter valid email address or try again later.");
						addFieldError("emailError",  errorMessage.toString());
						return INPUT;
					}

					// Insert only after email verification is sent
					userService.insertUser(newUser);

					// Email the admin to assign permissions for new user
					StringBuffer adminMessageBody = new StringBuffer();
					adminMessageBody.append("New account created in Metadata Services Toolkit with user name : " + newUser.getUsername().trim());
					adminMessageBody.append("\nComments from the user : " + comments);
					adminMessageBody.append("\nPlease login into the system and assign appropriate permissions for the user.");
					String adminSubject = "Assign permission to new User";

					// TODO Change to get all admin of system and email to their ids
					emailer.sendEmail(userService.getUserByUserName("admin", serverService.getServerByName("Local")).getEmail(), adminSubject, adminMessageBody.toString());
				} else {
					servers =  serverService.getAll();
					addFieldError("userEmailExist", "This email address already exist in the system.- " + newUser.getEmail().trim());
					return INPUT;
				}

			} else {
				servers =  serverService.getAll();
				addFieldError("userNameExist", "User name already exist - " + newUser.getUsername().trim());
				return INPUT;
			}
		} catch (Exception e) {
			addFieldError("dataError", e.getMessage());
			return INPUT;
		}
		return SUCCESS;
	}

	public User getNewUser() {
		return newUser;
	}

	public void setNewUser(User newUser) {
		this.newUser = newUser;
	}

	public List<Server> getServers() {
		return servers;
	}

	public String getComments() {
		return comments;
	}

	public void setComments(String comments) {
		this.comments = comments;
	}

	public String getServerName() {
		return serverName;
	}

	public void setServerName(String serverName) {
		this.serverName = serverName;
	}

}
